File: algo_ds/_general/merge_Sort.cpp
#include <iostream>
#include<bits/stdc++.h>
#include<vector>
using namespace std;

void sorter(vector<int>&arr,int low,int mid,int high){
	int start1=low,start2=mid+1;
	vector<int>temp;
		for(auto i:arr)
		cout<<i<<" ";
		cout<<endl;
	while(start1<=mid && start2 <=high)
	{	if(arr[start1]<arr[start2])
		{	temp.push_back(arr[start1]);
			start1+=1;
		}
		else
			{	temp.push_back(arr[start2]);
			start2+=1;
		}
	}

	while(start1<=mid)
	{	temp.push_back(arr[start1]);
		start1+=1;
	}
	while(start2 <=high)
	{	temp.push_back(arr[start2]);
		start2+=1;
	}
	int k=0;
	for(auto i=low;i<=high;i++)
		arr[i]=temp[k++];

	// Mistake when copying back copy arr[low]~arr[high]=temp[0]~temp[size]

	for(auto i:arr)
		cout<<i<<" ";

	cout<<" low: "<<low<<" mid "<<mid<<" high "<<high<<endl ;

}


void merge(vector<int>& arr,int low,int high)
{	if(low<high)
	{	int mid=(low+(high-1))/2;
		merge(arr,low,mid);
		merge(arr,mid+1,high);
		sorter(arr,low,mid,high);

	}
}


int main() {
	// your code goes here
	vector<int>arr={1,4,3,6,6,3,8,23,4};

	merge(arr,0,arr.size());
	for(auto i:arr)
		cout<<i;



	return 0;
}

File: scratch/algo_ds/_general/merge_sort.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Oct 12 20:13:25 2017

@author: gowtham
"""
def sorter(arr,low,mid,high):
    start1=low
    start2=mid+1
    temp=[]
    
    start1=int(start1)
    start2=int(start2)
    while(start1<=mid and start2<=high):
        if(arr[start1]<arr[start2]):
            temp.append(arr[start1])
            start1=start1+1
        else:
            temp.append(arr[start2])
            start2=start2+1
            
    while(start1<=mid):
        temp.append(arr[start1])
        start1=start1+1
        
    while(start2<=high):
        temp.append(arr[start2])
        start2=start2+1
        
    arr=temp        
        
    

def merge(l,low,high):
    if(int(low)<int(high)):
        mid=(low+high)/2
        merge(l,low,mid)
        merge(l,mid+1,high)
        sorter(l,low,mid,high)

if __name__=='main':
    l=[34,343,54,5,555,85]
    
else:
    l=[34,343,54,5,555,85]
    l.sort()
    merge(l,0,int(len(l)-1))
    print (l)